Jake Jsonparser step5

ChocOZerO opened the pull request:
JSON 문자열 생성 구현

ChocOZerO committed:
[Add]JSONPainter를 활용해 JSON 출력 양식 추가

ChocOZerO committed:
[Update]배열에따라 depth가 정의되는 방식으로 변경

godrm commented:
```
typealias JSONData = Array<Any>
typealias JSONObject = Dictionary<String, Any>
```
이 부분은 main에 있기보다는 다른 파일에 있어야 할 것 같습니다. 한참 찾았네요 :(
main 파일은 재사용되지 않는 코드입니다. 전체에서 사용하는 타입이니 관련있는 파일에 모아서 선언해놓는게 좋습니다.

ChocOZerO committed:
[FitCodacy]공백 수정

ChocOZerO committed:
[Update]typealias가 선언된 위치를 관련 파일로 이동

godrm commented on JSONParser/JSONParser/JSONPainter.swift:
```diff
@@ -27,7 +27,7 @@ struct JSONPainter {
         paintArrayType()
```
특별히 `jsonData.count == 1`인 경우만 처리는 이유가 있나요?
이 정도 if 비교문도 두 개로 나누기보다는 합쳐서 `if jsonData.count == 1, let jsonObject = jsonData[0] as? JSONObject {`형태는 어떨까요?

ChocOZerO replied:
배열에 객체 하나만 있을 경우에 “[”,“]”가 출력되지 않게 하려고 구분했습니다.
알려주신 방법으로 바꿔보겠습니다.

godrm replied:
단지 배열에 항목이 하나만 있을때를 구분하기 위한게 아니라 메소드 자체가 달라보였는데
실제로 이 부분이 없어도 paintInsideOfArray() 에서 `,`는 빼고 출력할 것 같아 보였어요. 아닌가요?  ㅎㅎ

godrm replied:
아.. 다시보니 빨간색 delete 한 코드였구나.. ㅜㅜ

ChocOZerO committed:
[Update]조건문 개선해서 조건을 명확하게 바꿈

JSONData에 객체 한개만 들어왔을 경우 {}로 감싸진 결과물을 출력
JSONData의 길이가 1이고 그 해당 데이터가 JSONObject일 경우만 해당

ChocOZerO committed:
[Refactoring]JSONData -> JSONType 프로토콜로 변경

JSONType을 채택하는 JSONArray와 JSONObject 활용

ChocOZerO committed:
[FitCodacy]타입 강제변환 제거

godrm commented on JSONParser/JSONParser/JSONArray.swift:
```diff
@@ -0,0 +1,24 @@
+
+struct JSONArray: JSONType {
+    private(set) var JSONData: Array<String>
+    private(set) var JSONArray: Array<Any>
```
변수명은 대문자로 쓰면 헷갈려서 실수할 수 있으니 소문자로 쓰세요.

godrm commented on JSONParser/JSONParser/TypeCounter.swift:
```diff
@@ -48,16 +45,16 @@ struct TypeCounter {
     private mutating func countObjectType(jsonObject: JSONObject) {
-        for item in jsonObject {
-            if item.value is JSONObject {
+        for value in jsonObject.JSONObject.values {
```
속성(변수)이름을 대분자로 쓰니까 `jsonObject.JSONObject.values` 형태가 어색해졌습니다.

godrm commented on JSONParser/JSONParser/JSONObject.swift:
```diff
@@ -0,0 +1,24 @@
+
+struct JSONObject: JSONType {
+    private(set) var JSONData: Array<String>
+    private(set) var JSONObject: Dictionary<String, Any>
```
아래 코드를 보다보니 차라리 JSONObject나 JSONData를 접근할 수 있도록 열기보다는
subscript 로 구현하는게 더 좋을것 같습니다. 어떻게 생각하나요?

ChocOZerO replied:
아 그 방법이 더 좋을 것 같네요
근데 프로토콜을 채택하기위해 JSONData를 만든건데 사실 불필요해서 다른방식으로 바꿔야할거같아요

ChocOZerO committed:
[Fix]배열과 객체 카운트 오류 해결

godrm commented on JSONParser/JSONParserTests/JSONObjectTest.swift:
```diff
@@ -0,0 +1,40 @@
+    func testAdd() {
+        jsonObject.add(keyValue: (key: "jake", value: 3))
+        XCTAssertEqual(jsonObject.JSONObject.count, 1)
+        XCTAssertEqual(jsonObject.JSONData.count, 1)
```
이 테스트코드를 보니, 혹시 JSONObject나 JSONArray를 DTO로 쓰려고 했던건가요?

ChocOZerO replied:
네 처음엔 카운트하는 프로퍼티도 다 넣었다가 너무 공사가 커질거 같아서 선회했습니다.

ChocOZerO committed:
[Refactoring]JSONArray와 JSONObject 타입 명확하게 개선

TypeCounter대신 각 객체에서 자신의 상태를 알 수 있도록 개선
그에따라 테스트 케이스와 각각 비교 로직 수정

godrm commented on JSONParser/JSONParser/JSONObject.swift:
```diff
@@ -0,0 +1,64 @@
+        countType(element: value)
+        count += 1
+        keys = [String](jsonObject.keys)
+        keyCount += 1
```
keyCount 라는게 꼭 있어야 하는 변수인가요? 
count와 keyCount는 결국 같은 값일것 같은데... 왜 별도로 만들었는지 설명해주세요.

ChocOZerO replied:
count의 경우는 객체에 새로운 키밸류 쌍이 추가될때마다 올라가서 JSONObect내에 항상 유지되는 크기값으로 고정시키면서 keyCount와의 차를 통해 인덱스 0부터 순서대로 출력하고 싶어서 이와같이 구현했습니다.

godrm replied:
바로 위에 keys.count 를 쓰면 안되는건가요?

ChocOZerO replied:
음 제가 defer에 대해 이해가 부족하여 쫄아서 방어적으로 코딩한거같아요.(defer 강의 요청드립니다. Sequence와 IterableProtocol 도.. ㅎㅎ)
다른 collection처럼 count라는 변수가 있어야될것 같다는 생각때문에 count를 만들고, count값이 떨어지지 않게 하고 싶어서 keyCount를 만들다 보니 keys.count를 쓸 생각을 아에 못했네요.

godrm replied:
이 부분만 바꾸면 될 것 같습니다. 개선해보세요 👍

ChocOZerO committed:
[FitCodacy]강제타입캐스팅 제거

ChocOZerO committed:
[Update]Iterable구현을 위한 변수 접근제어 및 이름 변경

ChocOZerO commented:
Iterable 관련 변수 수정 완료 했습니다.

godrm commented:
이번 경험이 다른 미션에서도 이어지길 기대하겠습니다. 수고하셨습니다.
